microgen/test_utils.py

```diff
@@ -0,0 +1,94 @@
+import numpy as np
+
+def is_periodic(nodes_coords, tol=1e-8, dim=3) -> bool:
+    # bounding box
+    xmax = np.max(nodes_coords[:, 0])
+    xmin = np.min(nodes_coords[:, 0])
+    ymax = np.max(nodes_coords[:, 1])
+    ymin = np.min(nodes_coords[:, 1])
+    if dim == 3:
+        zmax = np.max(nodes_coords[:, 2])
+        zmin = np.min(nodes_coords[:, 2])
+
+    # extract face nodes
+    left = np.where(np.abs(nodes_coords[:, 0] - xmin) < tol)[0]
+    right = np.where(np.abs(nodes_coords[:, 0] - xmax) < tol)[0]
+
+    if dim > 1:
+        bottom = np.where(np.abs(nodes_coords[:, 1] - ymin) < tol)[0]
+        top = np.where(np.abs(nodes_coords[:, 1] - ymax) < tol)[0]
+
+    if dim > 2:  # or dim == 3
+        back = np.where(np.abs(nodes_coords[:, 2] - zmin) < tol)[0]
+        front = np.where(np.abs(nodes_coords[:, 2] - zmax) < tol)[0]
+
+        # sort adjacent faces to ensure node correspondance
+    if nodes_coords.shape[1] == 2:  # 2D mesh
+        left = left[np.argsort(nodes_coords[left, 1])]
+        right = right[np.argsort(nodes_coords[right, 1])]
+        if dim > 1:
+            bottom = bottom[np.argsort(nodes_coords[bottom, 0])]
+            top = top[np.argsort(nodes_coords[top, 0])]
+
+    elif nodes_coords.shape[1] > 2:
+        decimal_round = int(-np.log10(tol) - 1)
+        left = left[
+            np.lexsort(
+                (nodes_coords[left, 1], nodes_coords[left, 2].round(decimal_round))
+            )
+        ]
+        right = right[
+            np.lexsort(
+                (nodes_coords[right, 1], nodes_coords[right, 2].round(decimal_round))
+            )
+        ]
+        if dim > 1:
+            bottom = bottom[
+                np.lexsort(
+                    (
+                        nodes_coords[bottom, 0],
+                        nodes_coords[bottom, 2].round(decimal_round),
+                    )
+                )
+            ]
+            top = top[
+                np.lexsort(
+                    (nodes_coords[top, 0], nodes_coords[top, 2].round(decimal_round))
+                )
+            ]
+        if dim > 2:
+            back = back[
+                np.lexsort(
+                    (nodes_coords[back, 0], nodes_coords[back, 1].round(decimal_round))
+                )
+            ]
+            front = front[
+                np.lexsort(
+                    (
+                        nodes_coords[front, 0],
+                        nodes_coords[front, 1].round(decimal_round),
+                    )
+                )
+            ]

```

Comment:
I think this part is hard to read due to the duplication, and it's difficult to follow what changes between the processing of the different dimensions. What about extracting a function? (even a local one would do fine) Something like:

```python
        def _sort_dim(indices: np.ndarray, dim_a: int, dim_b: int) -> np.ndarray:
            return indices[
                np.lexsort(
                    (nodes_coords[indices, dim_a], nodes_coords[indices, dim_b].round(decimal_round))
                )
            ]

        left = _sort_dim(left, dim_a=1, dim_b=2)
        right = _sort_dim(right, dim_a=1, dim_b=2)
        if dim > 1:
            bottom = _sort_dim(bottom, dim_a=0, dim_b=2)
            top = _sort_dim(top, dim_a=0, dim_b=2)
        if dim > 2:
            back = _sort_dim(back, dim_a=0, dim_b=1)
            front = _sort_dim(front, dim_a=0, dim_b=1)
```

Reply:
Thanks for the suggestion, this is not added